Max better pids

maxspier opened the pull request:
<h1> This is the new framework for PIDs. By the way, how do you do logs again? </h1>
For some reason, ya'll let me make it.
<br> It's pretty great.

<h2> How Has This Been Tested? </h2>
I didn't run any tests yet. (Ron respond to my discord messages please). 
<br> Because I'm so good at coding, it will probably work. Still need to implement the antigrav but thats some AP calc BC stuff so it isn't really for me.

<h3> Please take a look and let me know what you like/dislike </h3>
Thx, and as always, "the world belongs to those who prepare for it today" (Malcolm X).

maxspier committed:
test

maxspier committed:
i'm such a great developer

1yd1a commented on src/main/java/com/team766/controllers/CSMMCPID.java:
```diff
@@ -37,23 +39,24 @@ public class CSMMCPID {
 	public static double antiGrav1 = 0;
 	public static double currentPos = 0;
-	public static double combo = 0;
+	public static double combo;
```
You should probably instantiate this here to avoid error later

1yd1a commented:
You should definitely test this stuff before pulling it to main. Your changes seem ok other than the comments I made above, but you still need to make sure the code builds.

maxspier committed:
fixed log errors?

maxspier committed:
i swear to god if these logs dont work im finna remove them

maxspier committed:
removed logs because they aren't worth my time

jamesi8086 commented on src/main/java/com/team766/controllers/CSMMCPID.java:
```diff
@@ -0,0 +1,181 @@
+	// The attributes of the class include references to the motor controller, SparkMax controller, PID controller, and absolute encoder
+	private MotorController mc1;
+	private CANSparkMax csm1;
+	private SparkMaxPIDController pid1;
```
why is this `pid1` ? why not just `pid` or `smPid` (since its from a sparkmax)

maxspier replied:
because I made a mistake before and forgot object oriented programming was a thing. Will change back when oficial merge to maroon fw happens.

jamesi8086 reviewed (commented):
please use a java linter to fix formatting before merging
e.g.
https://code.visualstudio.com/docs/java/java-linting

jamesi8086 commented on src/main/java/com/team766/controllers/CSMMCPID.java:
```diff
@@ -0,0 +1,181 @@
+		mc1.setNeutralMode(mode);
+	}
+	//setting the maximum and minimul locations that the motor can go to
+	public void setMaxMinLocation(double max, double min){
```
I think the common convention is min before max

jamesi8086 commented on src/main/java/com/team766/controllers/CSMMCPID.java:
```diff
@@ -0,0 +1,181 @@
+	}
+
+	//run loop that actually runs the PID using normal dz
+	public void run(boolean enabled){
```
add a comment to suggest how often this needs to be called

maxspier replied:
fair, will do

jamesi8086 commented on src/main/java/com/team766/controllers/CSMMCPID.java:
```diff
@@ -0,0 +1,181 @@
+	public void run(boolean enabled){
+		if(enabled){
+			switch(theState){
+				case OFF:
```
you are missing a function to turn PID state to OFF / unpower the motor

jamesi8086 replied:
might also want to consider toggling motor brake seperately

maxspier replied:
done.
We have brake/coast toggaling already, no?

jamesi8086 commented on src/main/java/com/team766/controllers/CSMMCPID.java:
```diff
@@ -0,0 +1,181 @@
+		pid1.setReference(position, ControlType.kSmartMotion);
+	}
+	//1st step to go to a position using the normal PID, setting what you want the position to be
+	public void magicallyGoToPositionUsingNormalDZ(double position){
```
there must be a more acceptable function name than `magicallyGoToPositionUsingNormalDZ`

maxspier replied:
No thanks

maxspier replied:
Maybe later after fun testing

maxspier replied:
fine, done. setAllusion() because setReference() is Revs name

jamesi8086 commented on src/main/java/com/team766/controllers/CSMMCPID.java:
```diff
@@ -0,0 +1,181 @@
+				case ANTIGRAV:
+					//TODO: antigravity
+				case PID:
+					if (mc1.getSensorPosition() <= (dz1 + mc1.getSensorPosition()) && mc1.getSensorPosition() >= (mc1.getSensorPosition() - dz1)){
```
something looks incorrect with this line

1yd1a replied:
@TTVMixmix00 get henry to help you do antigrav stuff

maxspier replied:
asking them

1yd1a commented on src/main/java/com/team766/controllers/CSMMCPID.java:
```diff
@@ -0,0 +1,181 @@
+		pid1.setReference(position, ControlType.kSmartMotion);
+	}
+	//1st step to go to a position using the normal PID, setting what you want the position to be
+	public void magicallyGoToPositionUsingNormalDZ(double position){
```
just do "setPosition"...

maxspier replied:
No

maxspier replied:
see above conversation

maxspier commented:
> please use a java linter to fix formatting before merging e.g. https://code.visualstudio.com/docs/java/java-linting

still unsure what this is

maxspier committed:
updated new PID framework with new incorporations from PR

maxspier committed:
test changes - if they work im keeping, else removing

maxspier committed:
that was expected

maxspier committed:
done

maxspier committed:
i think it should be good now except all of the errors with the logging but idk what happened there it is all red

(/) Fixed conventional things
(+) Added try_catch + an exception on the constructor just incase someone mistypes a config name and we don't want the whole robot angry
(-) removed useless stuff
(/) Due to common vote, renamed setAllusion() to setPosition()

maxspier committed:
OI is so goofy

Why cant we just create a fork at the beginning of the season???

jamesi8086 commented on src/main/java/com/team766/controllers/CSMMCPID.java:
```diff
@@ -30,15 +31,15 @@ public class CSMMCPID{
 	private SparkMaxPIDController pid1;
 	private SparkMaxAbsoluteEncoder abs1;
 	//PID Related Variables (should be private but whatever)
```
now that they are private, remove this line

jamesi8086 commented on src/main/java/com/team766/controllers/CSMMCPID.java:
```diff
@@ -51,27 +52,36 @@ private enum PIDSTATE{
+				mc1 = RobotProvider.instance.getMotor(configName);
+				csm1 = (CANSparkMax)mc1;
+				pid1 = csm1.getPIDController();
+			}catch (IllegalArgumentException Ill){
```
variable names always begin lowercased

jamesi8086 commented on src/main/java/com/team766/controllers/CSMMCPID.java:
```diff
@@ -151,8 +155,8 @@ public void setMaxAccel(double max){
-	//go to a position using the thing that wasn't tested yet (ron respond to my discord messages please so i can test)
-	public void setCLEAllusion(double position){
+	//go to a position using the thing that wasn't tested yet and almost broke the robot...
+	public void setCLEPosition(double position){
```
What is a `CLEPosition` ?

maxspier replied:
CLE, or Closed Loop Error is Rev's built in deadzone function that we didn't know about until very recently (because they named the function to set it "setSmartMotionAllowedClosedLoopError()", which only after knowing what it did made sense). We haven't tested CLE on a robot yet.

jamesi8086 commented on src/main/java/com/team766/controllers/CSMMCPID.java:
```diff
@@ -173,6 +177,8 @@ public void setAllusion(double position){
 	}
 
 	public void STOP(){
```
function name should not be all caps

jamesi8086 reviewed (commented):
got a couple of things to fix there

maxspier committed:
patch 1a

maxspier committed:
patch 1b

maxspier committed:
patch 1c

maxspier reviewed (commented):
Thoughts now? (@yongkimleng)

jamesi8086 commented:
better, but the build is broken (see CI failures)
once you resolve that, its good

I hope @JayAgra has time to put in automatic linting into our github repos so that you/we can care less about manually conforming to Java code style, spacings, etc

maxspier committed:
found an error suprisingly
